File: VTablature.m
//
//  VTablature.m
//  tablution
//
//  Created by The Valrus on 9/30/08.
//  Copyright 2008 __MyCompanyName__. All rights reserved.
//

#import "VTablature.h"
#import "VChord.h"
#import "VNote.h"

#define MAX_STRINGS 7

@implementation VTablature

@synthesize chords;
@synthesize measureBars;
@synthesize numStrings;

NSString * const VTABLATURE_DATA_UTI = @"com.valrusware.tablature";
NSUInteger const HUMAN_READABLE_WIDTH = 80;

#pragma mark -
#pragma mark Init and setup

- (id)initWithStrings:(NSUInteger)num
{
    // Returns an initialized VTablature.
    self = [super init];
    if (self) {
        numStrings = 6; // TODO: make not hardcoded
        chords = [NSMutableArray arrayWithCapacity:10];
        measureBars = [NSMutableIndexSet indexSet];
        return self;
    } else {
        return nil;
    }
}

- (id)init
{
    NSLog(@"VTablature init");
    return [self initWithStrings:numStrings];
}

- (void)addFeaturesFromTextArray:(NSArray *)textArray
                         atIndex:(NSUInteger)index
{
    for (NSString *featureText in textArray) {
        if ([featureText isEqual:@"|"]) {
            // TODO: In general this shouldn't toggle
            [self toggleBarAtIndex:index];
        }
    }
}

+ (VTablature *)tablatureFromText:(NSString *)tabText
{
    VTablature *newTab = [[VTablature alloc] init];
    NSArray *chordTexts = [tabText componentsSeparatedByString:@"\n"];
    for (NSString *chordText in chordTexts) {
        if ([chordText length] == 0) {
            break;
        }
        NSArray *fretStringsArray = [chordText componentsSeparatedByString:@" "];
        if ([fretStringsArray count] >= [newTab numStrings]) {
            [newTab addFeaturesFromTextArray:[fretStringsArray subarrayWithRange:NSMakeRange([newTab numStrings],
                                                                                             [fretStringsArray count] - [newTab numStrings])]
                                     atIndex:[newTab countOfChords]];
        }
        [newTab addChordFromArray:[[fretStringsArray subarrayWithRange:NSMakeRange(0, [newTab numStrings])] valueForKey:@"intValue"]];
    }
    return newTab;
}

+ (VTablature *)tablatureWithChords:(NSArray *)chords
{
    VTablature *newTab = [[VTablature alloc] init];
    [newTab insertChords:[NSArray arrayWithArray:chords]
               atIndexes:[NSIndexSet indexSetWithIndexesInRange:NSMakeRange(0, [chords count])]];
    return newTab;
}

- (void)observeValueForKeyPath:(NSString *)keyPath
                      ofObject:(id)object
                        change:(NSDictionary *)change
                       context:(void *)context
{
    NSLog(@"VTablature sees a change in a chord's %@!", keyPath);
}

#pragma mark -
#pragma mark Teardown

- (void) dealloc
{
    [chords removeObserver:self
      fromObjectsAtIndexes:[NSIndexSet indexSetWithIndexesInRange:NSMakeRange(0, [chords count])]
                forKeyPath:@"notes"];
}

#pragma mark -
#pragma mark KVC-compliant accessors

- (id)objectInChordsAtIndex:(NSUInteger)index
{
    VChord *soughtChord;
    if ((soughtChord = [chords objectAtIndex:index])) {
        return soughtChord;
    } else {
        return nil;
    }
}

- (NSArray *)chordsAtIndexes:(NSIndexSet *)indexSet
{
    return [chords objectsAtIndexes:indexSet];
}

- (NSUInteger)countOfChords
{
    return [chords count]; 
}

- (void)insertObject:(VChord *)chord
     inChordsAtIndex:(NSUInteger)index
{
    [chords insertObject:chord atIndex:index];
    [chord addObserver:self forKeyPath:@"notes" options:0 context:NULL];
}

- (void)insertChords:(NSArray *)chordArray
           atIndexes:(NSIndexSet *)indexes
{
    [chords insertObjects:chordArray atIndexes:indexes];
    [chords addObserver:self
     toObjectsAtIndexes:indexes
             forKeyPath:@"notes"
                options:0
                context:NULL];
}

- (void)removeObjectFromChordsAtIndex:(NSUInteger)index
{
    if (index < [self countOfChords]) {
        [chords removeObserver:self
          fromObjectsAtIndexes:[NSIndexSet indexSetWithIndex:index]
                    forKeyPath:@"notes"
                       context:NULL];
        [chords removeObjectAtIndex:index];
    }
}

- (void)removeChordsAtIndexes:(NSIndexSet *)indexes
{
    [chords removeObserver:self
      fromObjectsAtIndexes:indexes
                forKeyPath:@"notes"
                   context:NULL];
    [chords removeObjectsAtIndexes:indexes];
}

- (void)replaceChordsAtIndexes:(NSIndexSet *)indexes
                    withChords:(NSArray *)array
{
    [self removeChordsAtIndexes:indexes];
    NSRange insertRange = NSMakeRange([indexes firstIndex], [array count]);
    [self insertChords:array
             atIndexes:[NSIndexSet indexSetWithIndexesInRange:insertRange]];
}

#pragma mark -
#pragma mark Other accessors

# pragma mark Convenience mutators

- (void)addChordFromArray:(NSArray *)chordArray
{
    VChord *newChord = [VChord chordWithArray:chordArray];
    [self insertObject:newChord inChordsAtIndex:[self countOfChords]];
}

- (void)addChordFromString:(NSString *)chordString
{
    VChord *newChord;
    if ((newChord = [VChord chordWithStrings:numStrings
                                    fromText:chordString])) {
        [self insertObject:newChord inChordsAtIndex:[self countOfChords]];
    }
    else {
        // invalid chord
    }
}

- (void)deleteNoteAtIndex:(NSUInteger)index
                 onString:(NSUInteger)stringNum
{
    [self insertNote:[VNote blankNote]
             atIndex:index onString:stringNum];
}

- (VChord *)lastChord
{
    return [self objectInChordsAtIndex:[self countOfChords] - 1];
}

- (VNote *)noteAtIndex:(NSUInteger)index
              onString:(NSUInteger)stringNum
{
    VChord *soughtChord;
    if ((soughtChord = [chords objectAtIndex:index])) {
        return [soughtChord objectInNotesAtIndex:stringNum];
    } else {
        return nil;
    }
}

- (NSInteger)fretAtIndex:(NSUInteger)index
                onString:(NSUInteger)stringNum
{
    VNote *soughtNote;
    if ((soughtNote = [self noteAtIndex:index onString:stringNum])) {
        return [soughtNote fret];
    } else {
        return -1;
    }
}

- (bool)hasBarAtIndex:(NSUInteger)index
{
    return ([measureBars containsIndex:index]);
}

#pragma mark Tab-level mutators

- (void)extend
{
    [self addChordFromString:@"-1 -1 -1 -1 -1 -1"];
}

- (void)toggleBarAtIndex:(NSUInteger)index
{
    [self willChangeValueForKey:@"bars"];
    if ([measureBars containsIndex:index]) {
        [measureBars removeIndex:index];
    }
    else {
        [measureBars addIndex:index];
    }
    [self didChangeValueForKey:@"bars"];
}

#pragma mark Note-level mutators

- (void)insertNoteAtIndex:(NSUInteger)index
                 onString:(NSUInteger)stringNum
                   onFret:(NSUInteger)fretNum
{
    [self insertNote:[VNote noteAtFret:fretNum]
             atIndex:index
            onString:stringNum];
}

- (void)insertNote:(VNote *)note
           atIndex:(NSUInteger)index
          onString:(NSUInteger)stringNum
{
    [chords removeObserver:self
      fromObjectsAtIndexes:[NSIndexSet indexSetWithIndex:index]
                forKeyPath:@"notes"];

    [self willChange:NSKeyValueChangeReplacement
     valuesAtIndexes:[NSIndexSet indexSetWithIndex:index]
              forKey:@"chords"];
    [[chords objectAtIndex:index] replaceObjectInNotesAtIndex:stringNum
                                                   withObject:note];
    [self didChange:NSKeyValueChangeReplacement
    valuesAtIndexes:[NSIndexSet indexSetWithIndex:index]
             forKey:@"chords"];

    [chords addObserver:self
     toObjectsAtIndexes:[NSIndexSet indexSetWithIndex:index]
             forKeyPath:@"notes"
                options:0
                context:NULL];
}

- (void)insertChordFromArray:(NSArray *)chordArray
                     atIndex:(NSUInteger)index
{
    VChord *newChord = [VChord chordWithArray:chordArray];
    [self insertObject:newChord
       inChordsAtIndex:index];
}

#pragma mark -
#pragma mark Converting to strings

+ (NSString *)getNoteTextForString:(NSString *)fretText
{
    // A note with a string marked should look like "-2-" or "-13"
    // depending on the length of the fret number. Prepend a hyphen
    // and then append enough more to make the total length 5.
    return [@"-" stringByAppendingString:[fretText stringByPaddingToLength:2 
																withString:@"—"
														   startingAtIndex:0]];
}

+ (NSString *)getNoteTextForValue:(NSInteger)fretNum
{
    return [self getNoteTextForString:(fretNum >= 0
                                       ? [NSString stringWithFormat:@"%lu", fretNum]
                                       : @"-")];
}

- (NSString *)toSerialString
{
    NSMutableString *tabString = [NSMutableString string];
    for (NSUInteger chordIndex = 0; chordIndex < [self countOfChords]; ++chordIndex) {
        [tabString appendString:[[self objectInChordsAtIndex:chordIndex] asText]];
        if ([self hasBarAtIndex:chordIndex]) {
            [tabString appendString:@" |"];
        }
        [tabString appendString:@"\n"];
    }
    return [NSString stringWithString:tabString];
}

- (NSString *)toHumanReadableString
{
    NSMutableString *tabString = [NSMutableString string];
    NSMutableString *oneLineString = [NSMutableString stringWithCapacity:81];
    NSUInteger chordIndex = 0;
    NSUInteger startingIndexForThisRow = 0;
    NSUInteger stringIndex = 0;
    while ((chordIndex < [self countOfChords]) && (stringIndex < [self numStrings])) {
        NSString *nextPiece = [VTablature getNoteTextForValue:[self fretAtIndex:chordIndex
                                                                       onString:stringIndex]];
        if ([oneLineString length] + [nextPiece length] < HUMAN_READABLE_WIDTH) {
            [oneLineString appendString:nextPiece];
            chordIndex++;
        }
        else {
            [oneLineString appendString:@"\n"];
            [tabString appendString:oneLineString];
            [oneLineString setString:@""];
            if (stringIndex + 1 < [self numStrings]) {
                // go to next string, "carriage return" for chord index
                stringIndex++;
                chordIndex = startingIndexForThisRow;
            }
            else if (chordIndex < [self countOfChords]) {
                // next line of tab
                [tabString appendString:@"\n"];
                startingIndexForThisRow = chordIndex;
                stringIndex = 0;
            }
        }
    }
    return tabString;
}

- (NSArray *)asArrayOfStrings
{
    return [chords valueForKey:@"asText"];
}

#pragma mark -
#pragma mark NSFastEnumeration protocol

- (NSUInteger)countByEnumeratingWithState:(NSFastEnumerationState *)state
                                  objects:(id __unsafe_unretained [])stackbuf
                                    count:(NSUInteger)len
{
    // just "delegate" this to internal NSMutableArray
    return [chords countByEnumeratingWithState:state objects:stackbuf count:len];
}

#pragma mark -
#pragma mark NSPasteboardWriting protocol

- (NSArray *)writableTypesForPasteboard:(NSPasteboard *)pasteboard
{
    return [NSArray arrayWithObject:VTABLATURE_DATA_UTI];
}

- (NSPasteboardWritingOptions)writingOptionsForType:(NSString *)type
                                         pasteboard:(NSPasteboard *)pasteboard
{
    return 0;
}

- (id)pasteboardPropertyListForType:(NSString *)type
{
    if ([type isEqualToString:VTABLATURE_DATA_UTI]) {
        return [self toSerialString];
    }
    return nil;
}

#pragma mark -
#pragma mark NSPasteboardReading protocol

+ (NSArray *)readableTypesForPasteboard:(NSPasteboard *)pasteboard
{
    return [NSArray arrayWithObject:VTABLATURE_DATA_UTI];
}

+ (NSPasteboardReadingOptions)readingOptionsForType:(NSString *)type
                                         pasteboard:(NSPasteboard *)pasteboard
{
    if ([type isEqualToString:VTABLATURE_DATA_UTI]) {
        return NSPasteboardReadingAsString;
    }
    return 0;
}

- (id)initWithPasteboardPropertyList:(id)propertyList
                              ofType:(NSString *)type
{
    if ([type isEqualToString:VTABLATURE_DATA_UTI]) {
        return [VTablature tablatureFromText:propertyList];
    }
    return nil;
}

#pragma mark -
#pragma mark NSCoding protocol

#define kChords     @"chords"
#define kStrings    @"numStrings"

- (void)encodeWithCoder:(NSCoder *)coder {
    [coder encodeObject:[self chords] forKey:kChords];
}

@end

File: VTablature.swift
//
//  VTablature.swift
//  tablution
//
//  Created by Ian McCowan on 11/29/14.
//
//

import AppKit
import Foundation

let HUMAN_READABLE_WIDTH = 80
let VTABLATURE_DATA_UTI = "com.valrusware.tablature"

public struct TabLocation {
    var index = 0
    var string = 0
}

@objc public class VTablature: NSObject, SequenceType, NSPasteboardWriting, NSPasteboardReading {
    dynamic var chords: Array<VChord>
    private var measureBars: NSMutableIndexSet
    
    var numStrings: Int
    var stringLabels: [String]
    
    // https://developer.apple.com/library/ios/documentation/Swift/Conceptual/BuildingCocoaApps/AdoptingCocoaDesignPatterns.html
    private var myContext = 0

    // MARK: Initializers and deinitializers
    
    init(numStrings: Int = 6, withChords chords: [VChord], withBarsAtIndexes barIndexes: NSMutableIndexSet) {
        self.numStrings = numStrings
        self.stringLabels = ["E", "A", "D", "G", "B", "e"]
        self.chords = chords
        self.measureBars = barIndexes
    }
    
    public convenience init(numStrings: Int) {
        self.init(numStrings: numStrings, withChords:[], withBarsAtIndexes:NSMutableIndexSet())
    }
    
    deinit {
        for chord in self.chords {
            chord.removeObserver(self, forKeyPath: "notes")
        }
    }
    
    func addFeaturesFromTextArray(textArray: Array<String>, atIndex index: Int)
    {
        for featureText in textArray {
            if featureText == "|" {
                self.toggleBarAtIndex(index)
            }
        }
    }

    convenience init(fromText tabText: String) {
        var numStrings: Int = 0
        let chordTextArray: Array<String> = tabText.characters.split(isSeparator: { $0 == "\n" }).map { String($0) }
        var chords: Array<VChord> = []
        let barIndexes: NSMutableIndexSet = NSMutableIndexSet()
        for chordText in chordTextArray {
            if chordText.isEmpty {
                break
            }
            var fretStringsArray = chordText.characters.split(isSeparator: { $0 == " " }).map { String($0) }
            if numStrings == 0 {
                // initialize a new tab with the number of strings seen
                numStrings = fretStringsArray.count
            }
            if fretStringsArray.count >= numStrings {
                let stringArray: [String] = Array(fretStringsArray[0..<numStrings])
                let featureArray = Array(fretStringsArray[numStrings..<fretStringsArray.count])
                if featureArray.contains("|") {
                    barIndexes.addIndex(chords.count)
                }
                chords.append(VChord.chordWithIntArray(stringArray.map({
                    (fretStr: String) -> Int in
                    if let fretNum = Int(fretStr) {
                        return fretNum
                    } else {
                        return -1
                    }
                })))
            }
        }
        self.init(numStrings: numStrings, withChords: chords, withBarsAtIndexes: barIndexes)
    }
    
    // MARK: Class functions for initialization
    
    public class func tablatureFromText(tabText: String) -> VTablature {
        return VTablature(fromText: tabText)
    }
    
    class func tablatureWithChords(chordArray: Array<VChord>) -> VTablature {
        if chordArray.isEmpty {
            return VTablature(numStrings: 6)
        }
        else {
            let newTablature = VTablature(numStrings: chordArray[0].numStrings())
            let insertionIndexSet = NSIndexSet(indexesInRange: NSRange(location:0,
                                                                       length: chordArray.count))
            newTablature.insertChords(chordArray, atIndexes:insertionIndexSet)
            return newTablature
        }
    }
    
    // MARK: - Accessors -
    
    public func objectInChordsAtIndex(index: Int) -> AnyObject {
        return self.chords[index]
    }

    public func chordsAtIndexes(indexes: NSIndexSet) -> Array<VChord> {
        return (self.chords as NSArray).objectsAtIndexes(indexes) as! Array<VChord>
    }
    
    public func countOfChords() -> Int {
        return chords.count
    }
    
    public func hasBarAtIndex(index: Int) -> Bool {
        return measureBars.containsIndex(index)
    }
    
    public func noteAtLocation(loc: TabLocation) -> VNote? {
        if loc.index < self.chords.count && loc.string < self.numStrings {
            return self.chords[loc.index][loc.string]
        }
        return nil
    }
    
    public func fretAtIndex(index: Int, onString stringNum: Int) -> Int? {
        if let note = self.noteAtLocation(TabLocation(index: index, string: stringNum)) {
            return note.fret
        }
        return nil
    }
    
    // String conversions
    
    class func getNoteTextForString(fretText: String) -> String {
        // A note with a string marked should look like "-2-" or "-13"
        // depending on the length of the fret number. Prepend a hyphen
        // and then append enough more to make the total length 3.
        return "-".stringByAppendingString(fretText.stringByPaddingToLength(2, withString: "-", startingAtIndex: 0))
    }
    
    class func getNoteTextForValue(fretNum: Int) -> String {
        return self.getNoteTextForString(fretNum >= 0 ? String(format: "%lu", fretNum) : "-")
    }
    
    func toSerialString() -> String {
        return Array(self.chords.enumerate()).map({
            String(format: "%@%@\n", (self.hasBarAtIndex($0) ? " |" : ""), $1.asText())
        }).joinWithSeparator("")
    }
    
    func toHumanReadableString() -> String {
        var tabString = String()
        var oneLineString = String()
        var chordIndex = 0
        var startingIndexForThisRow = 0
        var stringIndex = 0
        while chordIndex < self.chords.count && stringIndex < self.numStrings {
            let nextPiece: String = VTablature.getNoteTextForValue(self.fretAtIndex(chordIndex, onString: stringIndex)!)
            if (oneLineString.characters.count + nextPiece.characters.count < HUMAN_READABLE_WIDTH ) {
                oneLineString += nextPiece
                chordIndex += 1
            }
            else {
                oneLineString += "\n"
                tabString += oneLineString
                if stringIndex + 1 < self.numStrings {
                    // go to next string, "carriage return" for chord index
                    stringIndex += 1
                    chordIndex = startingIndexForThisRow
                }
                else if chordIndex < self.chords.count {
                    // next line of tab
                    tabString += "\n"
                    startingIndexForThisRow = chordIndex
                    stringIndex = 0
                }
            }
        }
        return tabString
    }
    
    // Iteration
    
    public func generate() -> NSFastGenerator {
        return NSFastGenerator(self.chords)
    }
    
    
    // MARK: - Mutators -
    
    // MARK: KVC/KVO and "base" mutators
    
    override public class func automaticallyNotifiesObserversForKey(key: String) -> Bool {
        return false
    }
    
    override public func observeValueForKeyPath(keyPath: String?, ofObject object: AnyObject?, change: [String : AnyObject]?, context: UnsafeMutablePointer<Void>) {
        if context == &myContext {
            NSLog("VTablature sees a change in a chord's %@!", keyPath!)
        } else {
            super.observeValueForKeyPath(keyPath, ofObject: object, change: change, context: context)
        }
    }
    
    public func insertChord(chord: VChord, atIndex index: Int) {
        self.willChange(NSKeyValueChange.Insertion, valuesAtIndexes: NSIndexSet(index: index), forKey: "chords")
        self.chords.insert(chord, atIndex: index)
        self.didChange(NSKeyValueChange.Insertion, valuesAtIndexes: NSIndexSet(index: index), forKey: "chords")
        chord.addObserver(self, forKeyPath: "notes", options: NSKeyValueObservingOptions(), context: &myContext)
    }
    
    public func insertChords(chords: [VChord], atIndexes indexes: NSIndexSet, andNotify notify: Bool = true) {
        var chordIndex : Int = 0
        if notify {
            self.willChange(NSKeyValueChange.Insertion, valuesAtIndexes: indexes, forKey: "chords")
        }
        indexes.enumerateIndexesWithOptions(NSEnumerationOptions(), usingBlock: {
            (index: Int, stop: UnsafeMutablePointer<ObjCBool>) -> Void in
            self.chords.insert(chords[chordIndex], atIndex: index)
            chords[chordIndex].addObserver(self, forKeyPath: "notes", options: NSKeyValueObservingOptions(), context: &self.myContext)
            chordIndex += 1
        })
        if notify {
            self.didChange(NSKeyValueChange.Insertion, valuesAtIndexes: indexes, forKey: "chords")
        }
    }
    
    public func removeChordAtIndex(index: Int) {
        self.willChange(NSKeyValueChange.Removal, valuesAtIndexes: NSIndexSet(index: index), forKey: "chords")
        self.chords[index].removeObserver(self, forKeyPath: "notes")
        self.chords.removeAtIndex(index)
        self.didChange(NSKeyValueChange.Removal, valuesAtIndexes: NSIndexSet(index: index), forKey: "chords")
    }
    
    public func removeChordsAtIndexes(indexes: NSIndexSet, andNotify notify: Bool = true) {
        if notify {
            self.willChange(NSKeyValueChange.Removal, valuesAtIndexes: indexes, forKey: "chords")
        }
        indexes.sort(>).forEach({
            self.chords[$0].removeObserver(self, forKeyPath: "notes")
            self.chords.removeAtIndex($0)
        })
        if notify {
            self.didChange(NSKeyValueChange.Removal, valuesAtIndexes: indexes, forKey: "chords")
        }
    }
    
    public func replaceChordsAtIndexes(indexes: NSIndexSet, withChords chords: [VChord]) {
        self.willChange(NSKeyValueChange.Replacement, valuesAtIndexes: indexes, forKey: "chords")
        self.removeChordsAtIndexes(indexes, andNotify: false)
        let indexRange: NSRange = NSMakeRange(indexes.firstIndex, indexes.count)
        self.insertChords(chords, atIndexes:(NSIndexSet(indexesInRange: indexRange)), andNotify: false)
        self.didChange(NSKeyValueChange.Replacement, valuesAtIndexes: indexes, forKey: "chords")
    }
    
    func insertNote(note: VNote, atLocation loc: TabLocation) {
        self.chords[loc.index].removeObserver(self, forKeyPath: "notes")
        self.willChange(NSKeyValueChange.Replacement, valuesAtIndexes: NSIndexSet(index: loc.index), forKey: "chords")
        self.chords[loc.index][loc.string] = note
        self.didChange(NSKeyValueChange.Replacement, valuesAtIndexes: NSIndexSet(index: loc.index), forKey: "chords")
        self.chords[loc.index].addObserver(self, forKeyPath: "notes", options: NSKeyValueObservingOptions(), context: &myContext)
    }
    
    // MARK: Auxiliary mutators
    
    func addChordFromIntArray(intArray: Array<Int>) {
        self.insertChord(VChord.chordWithIntArray(intArray), atIndex: self.chords.count)
    }
    
    func addChordFromString(str: String) {
        self.insertChord(VChord.chordWithStrings(self.numStrings, fromText: str)!, atIndex: self.chords.count)
    }
    
    func insertFret(fret: Int, atIndex index: Int, onString stringNum: Int) {
        self.insertNote(VNote(fret: fret), atLocation: TabLocation(index: index, string: stringNum))
    }
    
    func deleteNoteAtIndex(index: Int, onString stringNum: Int) {
        self.insertNote(VNote.blankNote(), atLocation: TabLocation(index: index, string: stringNum))
    }

    func insertChordFromText(chordText: String, atIndex index: Int) {
        if let newChord: VChord = VChord.chordWithStrings(self.numStrings, fromText: chordText) {
            self.insertChord(newChord, atIndex: index)
        }
    }
    
    func extend() {
        self.addChordFromIntArray([Int](count: self.numStrings, repeatedValue: VNote.NO_FRET()))
    }
    
    func toggleBarAtIndex(index: Int) {
        self.willChangeValueForKey("measureBars")
        if self.measureBars.containsIndex(index) {
            measureBars.removeIndex(index)
        } else {
            measureBars.addIndex(index)
        }
        self.didChangeValueForKey("measureBars")
    }
    
    // MARK: - Pasteboard -
    
    public func writableTypesForPasteboard(pasteboard: NSPasteboard) -> [String] {
        return [VTABLATURE_DATA_UTI]
    }
    
    public func writingOptionsForType(type: String, pasteboard: NSPasteboard) -> NSPasteboardWritingOptions {
        return NSPasteboardWritingOptions()
    }
    
    public func pasteboardPropertyListForType(type: String) -> AnyObject? {
        if type == VTABLATURE_DATA_UTI {
            return self.toSerialString()
        }
        return nil
    }

    public class func readableTypesForPasteboard(pasteboard: NSPasteboard) -> [String] {
        return [VTABLATURE_DATA_UTI]
    }
    
    public class func readingOptionsForType(type: String, pasteboard: NSPasteboard) -> NSPasteboardReadingOptions {
        if type == VTABLATURE_DATA_UTI {
            return NSPasteboardReadingOptions.AsString
        }
        return NSPasteboardReadingOptions()
    }
    
    convenience required public init?(pasteboardPropertyList propertyList: AnyObject, ofType type: String) {
        if type != VTABLATURE_DATA_UTI {
            return nil
        }
        guard let tabText = propertyList as? String else {
            return nil
        }
        
        self.init(fromText: tabText)
    }
    
    public func encodeWithCoder(encoder: NSCoder) {
        encoder.encodeObject(self.chords, forKey:"chords")
    }
}